tests/tools/fio/workload.go

```diff
@@ -42,7 +45,156 @@ func (fr *Runner) writeFiles(fullPath string, opt Options) error {
 	return os.RemoveAll(filepath.Join(fr.LocalDataDir, relDirPath))
 }
+
+// DeleteDirAtDepth deletes a random directory at the given depth.
+func (fr *Runner) DeleteDirAtDepth(relBasePath string, depth int) error {
+	if depth == 0 {
+		return ErrCanNotDeleteRoot
+	}
+
+	fullBasePath := filepath.Join(fr.LocalDataDir, relBasePath)
+
+	return fr.operateAtDepth(fullBasePath, depth, os.RemoveAll)
+}
+
+// DeleteContentsAtDepth deletes some or all of the contents of a directory
+// at the provided depths.
+func (fr *Runner) DeleteContentsAtDepth(relBasePath string, depth, pcnt int) error {
+	fullBasePath := filepath.Join(fr.LocalDataDir, relBasePath)
+
+	return fr.operateAtDepth(fullBasePath, depth, func(dirPath string) error {
+		fileInfoList, err := ioutil.ReadDir(dirPath)
+		if err != nil {
+			return err
+		}
+
+		for _, fi := range fileInfoList {
+			const hundred = 100
+			if rand.Intn(hundred) < pcnt { // nolint:gosec
+				path := filepath.Join(dirPath, fi.Name())
+				err = os.RemoveAll(path)
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		return nil
+	})
+}
+
+// List of known errors.
+var (
+	ErrNoDirFound       = errors.New("no directory found at this depth")
+	ErrCanNotDeleteRoot = errors.New("can not delete root directory")
+)
+
+func (fr *Runner) operateAtDepth(path string, depth int, f func(string) error) error {
+	if depth <= 0 {
+		log.Printf("performing operation on directory %s\n", path)
+		return f(path)
+	}
+
+	fileInfoList, err := ioutil.ReadDir(path)
+	if err != nil {
+		return errors.Wrapf(err, "unable to read dir at path %v", path)
+	}
+
+	var dirList []string
+
+	for _, fi := range fileInfoList {
+		if fi.IsDir() {
+			dirList = append(dirList, filepath.Join(path, fi.Name()))
+		}
+	}
+
+	rand.Shuffle(len(dirList), func(i, j int) {
+		dirList[i], dirList[j] = dirList[j], dirList[i]
+	})
+
+	for _, dirName := range dirList {
+		err = fr.operateAtDepth(dirName, depth-1, f)
+		if !errors.Is(err, ErrNoDirFound) {
+			return err
+		}
+	}
+
+	return ErrNoDirFound
+}
+
+func (fr *Runner) writeFilesAtDepth(fromDirPath string, depth, branchDepth int, opt Options) error {
+	if depth <= 0 {
+		return fr.writeFiles(fromDirPath, opt)
+	}
+
+	var subdirPath string
+
+	if branchDepth > 0 {
+		subdirPath = pickRandSubdirPath(fromDirPath)
+	}
+
+	if subdirPath == "" {
+		var err error
+
+		// Couldn't find a subdir, create one instead
+		subdirPath, err = ioutil.TempDir(fromDirPath, "dir_")
+		if err != nil {
+			return errors.Wrapf(err, "unable to create temp dir at %v", fromDirPath)
+		}
+	}
+
+	return fr.writeFilesAtDepth(subdirPath, depth-1, branchDepth-1, opt)
+}
+
+func pickRandSubdirPath(dirPath string) (subdirPath string) {
+	subdirCount := 0
+
+	fileInfoList, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		return ""
+	}
+
+	for _, fi := range fileInfoList {
+		if fi.IsDir() {
+			subdirCount++
+
+			// Decide if this directory will be selected - probability of
+			// being selected is uniform across all subdirs
```

Comment:
No it doesn't break the loop after the first subdirectory. 

It's a variation on [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling). I was going for selecting a single subdirectory among an arbitrarily large number of elements that may or may not be directories, in a single pass through the list, without a need to store an unknown number of subdirectory paths/indices in memory for later selection.

Essentially the first subdirectory we find has a 100% chance of being picked if there is only one subdirectory.
If we find a second subdirectory, there's a 50% chance of selecting that one instead, if there are only two, leaving a 50% chance of sticking with the first one we found.
If there is a third, we'll have a 1/3 chance of returning that one, or a 2/3 chance of returning whatever had already been resolved with uniform probability between dirs 1 and 2. Thus each of the 3 subdirectories has a uniform 1/3 probability of being selected.
And so on.

The cost is a random number roll for every subdirectory. Swapping to a memory-centric algorithm is trivial though; no qualms with making the switch; should be the same outcome. What are your thoughts? Any preference? More clarifying comments needed obviously if we stick with this algorithm :)

Reply:
Got it, makes sense now. I'm just thick.

Reply:
Followup, and I think this may be partly what threw me off.

The entire directory is read upfront, so at least the memory cost of reading the entire directory and keeping the names is memory has already been paid. Moreover, `fileInfoList` is sorted alphabetically.

Now, `fileInfoList` contains both directories and files, so there's no way of getting the subset of directories without going through all the file infos, and that's what this is doing.

The other part that is not completely clear to me is whether the file infos are read and loaded upfront, or whether they are "lazily" loaded as they are accessed, e.g., when `fi.IsDir()`. We can look at the implementation, but since it is not explicitly documented, the safest thing to assume is that all the file infos are loaded upfront. Then, if that's the case, all the memory cost is already incurred.

Reply:
Yes you're right that the `ReadDir` pulls the info in to memory already. I suppose this method was to prevent the need to put together another arbitrary sized data structure to track which entries are directories (and therefore random candidates), when a single value was all we're looking for. That said, it would have been a minor cost, since the size of that extra data would be less than the size of the file info list anyway. :shrug:

Reply:
BTW, no change requested. Merely trying to understand the tradeoffs.